### PR TITLE
Add tests for hierarchy caching functionality

### PR DESCRIPTION
**Pull request by edmofro**
And update functionality to meet the tests - should've done TDD from the start!

I had missed the fact that due to the forking recursive function, it wasn't deciding whether to use canonical or entity relation links for a whole generation, but just for each specific relationship. Tests caught, so refactored to make the recursion less forky (which also means we don't need a task queue for batching)

### CODE REVIEW

File: packages/database/src/testUtilities/populateTestData.js [hunk 1]
```diff
@@ -16,8 +16,8 @@ export const populateTestData = async (models, recordsByModelType) => {
   // process sequentially, as some inserts may depend on earlier foreign keys being inserted
   for (let i = 0; i < modelTypes.length; i++) {
     const modelType = modelTypes[i];
-    await Promise.all(
-      recordsByModelType[modelType].map(record => upsertDummyRecord(models[modelType], record)),
-    );
+    for (const record of recordsByModelType[modelType]) {
+      await upsertDummyRecord(models[modelType], record);
```
Comment (kael89) [inline packages/database/src/testUtilities/populateTestData.js:20]: 🎉 for `for ... of`!

File: packages/database/src/tests/cachers/EntityHierarchyCacher.test.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  EXPECTED_INITIAL_ANCESTOR_DESCENDANT_RELATIONS,
+} from './EntityHierarchyCacher.fixtures';
+
+describe('EntityHierarchyCacher', () => {
+  const models = getTestModels();
+  const hierarchyCacher = new EntityHierarchyCacher(models);
+
+  before(async () => {
+    const testData = await getTestData(models);
+    await populateTestData(models, testData);
```
Comment (kael89) [inline packages/database/src/tests/cachers/EntityHierarchyCacher.test.js:22]: So clean!! 🥇

File: packages/database/src/cachers/EntityHierarchyCacher.js [hunk 3]
```diff
@@ -28,64 +26,77 @@ export class EntityHierarchyCacher {
...
+      parent_id: entityIds,
       entity_hierarchy_id: hierarchyId,
     });
-    const childIds = hierarchyLinks.map(l => l.child_id);
-
-    if (childIds.length > 0) {
-      return childIds;
-    }
+    // if no entity relation links, the next gen must use the canonical "parent_id" links
+    return hierarchyLinks.length === 0;
```
Comment (kael89) [inline packages/database/src/cachers/EntityHierarchyCacher.js:96]: Using `.count` here could probably make this request faster

**Review by kael89 — CHANGES_REQUESTED** [review]
Great tests!! They will be very valuable in any update of this critical functionality 💯 

The fixture logic is very readable as is, thanks to your comments. Given serial alphabetical letters are used, I would be tempted to abstract their creation (someting of the sort: `createEntity(startLetter, endLetter)`). This may result in them being even easier to read, on the other hand there's a risk for the opposite to happen.

A couple of minor changes requested

File: packages/database/src/cachers/EntityHierarchyCacher.js (hunk 3 above)
Comment (edmofro) [reply packages/database/src/cachers/EntityHierarchyCacher.js:96]: True!

File: packages/database/src/cachers/EntityHierarchyCacher.js [hunk 4]
```diff
@@ -28,64 +26,77 @@ export class EntityHierarchyCacher {
...
-    const alreadyCached = !!(await this.models.ancestorDescendantRelation.findOne({
+    const numberAlreadyCached = await this.models.ancestorDescendantRelation.count({
       entity_hierarchy_id: hierarchyId,
-      ancestor_id: parentId,
-    }));
-    if (alreadyCached) {
+      ancestor_id: parentIds,
+      generational_distance: 1,
+    });
+    if (numberAlreadyCached === parentIds.length) {
```
[the remark this answers is not in the document]
Comment (edmofro) [reply packages/database/src/cachers/EntityHierarchyCacher.js:54]: Good call

File: packages/database/src/cachers/EntityHierarchyCacher.js [hunk 5]
```diff
@@ -28,64 +26,84 @@ export class EntityHierarchyCacher {
...
+    await this.fetchAndCacheDescendants(hierarchyId, childIdsToAncestorIds);
   }
 
-  async getNextGeneration(hierarchyId, entityId) {
-    // get any matching alternative hierarchy relationships leading out of these parents
-    const hierarchyLinks = await this.models.entityRelation.find({
-      parent_id: entityId,
+  async checkShouldFetchAndCache(hierarchyId, parentIds) {
+    if (parentIds.length === 0) {
+      return false; // base case of recursion, we must have reached the leaf nodes
```
Comment (kael89) [inline packages/database/src/cachers/EntityHierarchyCacher.js:83]: Perhaps the reference to recursion doesn't make sense now, but I'll leave this decision to you!

**Review by kael89 — APPROVED** [review]

**Comment by kael89** [comment]
One seemingly unrelated test case failed... Weird...

File: packages/database/src/cachers/EntityHierarchyCacher.js (hunk 5 above)
Comment (edmofro) [reply packages/database/src/cachers/EntityHierarchyCacher.js:83]: I still find it useful :-)

**Comment by edmofro** [comment]
Have restarted the build...we have a few flaky tests lately

**Pull request merged**